### PR TITLE
Issue/handle login in webview

### LINKED ISSUE #162: Handle login in webview
**Is your feature request related to a problem? Please describe.**
To better handle the login for different distributions we should load the `schul-cloud.org/login` page in a webview and fetch the authentication token from there.

### PR DESCRIPTION
**Pull request by ctiedt**
**Fixes: #162**

### CODE REVIEW

File: lib/sign_in/widgets/form.dart [hunk 1]
```diff
@@ -2,41 +2,32 @@ import 'package:flutter/material.dart';
...
   bool _isLoading = false;
   String _ambientError;
 
-  Future<void> _executeSignIn(Future<void> Function() signIn) async {
+  Future<void> _executeLogin(Future<void> Function() login) async {
     setState(() => _isLoading = true);
 
     try {
-      await signIn();
+      await login();
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: These and related functions were renamed to use "signIn" instead of "login" in their names recently, but probably the merge went wrong. Can you please adopt the new naming?

File: lib/sign_in/widgets/form.dart [hunk 2]
```diff
@@ -82,48 +67,62 @@ class _SignInFormState extends State<SignInForm> {
...
+
+                    final storage = services.get<StorageService>();
+                    await storage.token.setValue(jwt);
+
+                    unawaited(context.navigator.pushReplacement(
+                        TopLevelPageRoute(builder: (_) => SignedInScreen())));
+                  }
+                },
+                initialUrl:
+                    'https://${services.get<AppConfig>().domain}/login'),
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: - slight formatting change
- use `services.storage` i/o `services.get<StorageService>()`
- only test for path segment of URI
- use `AppConfig.webUrl` to build `initialUrl`
```suggestion
              javascriptMode: JavascriptMode.unrestricted,
              onWebViewCreated: (controller) => this.controller = controller,
              onPageFinished: (url) async {
                final firstPathSegment = Uri.parse(url).pathSegments.first;
                if (firstPathSegment == 'login') {
                  // The JavaScript is meant to isolate the login section.
                  // Hopefully there will be an option to only get that in a
                  // non-hacky way in the future.
                  await controller.evaluateJavascript('''
                  var node = document.getElementById('loginarea');
                  var html = document.getElementsByTagName('html')[0];
                  html.removeChild(html.childNodes[2]);
                  html.appendChild(node);
                  document.getElementsByTagName('h2')[0].innerHTML = ''
                  ''');
                } else if (firstPathSegment == 'dashboard') {
                  final cookies =
                      await controller.evaluateJavascript('document.cookie');
                  // Yes, this is not elegant. You may complain about it when
                  // there is a nice way to get a single cookie via JavaScript.
                  final jwt = cookies
                      .split('; ')
                      .firstWhere((element) => element.startsWith('"jwt='))
                      .replaceAll('"', '')
                      .substring(4);

                  await services.storage.token.setValue(jwt);

                  unawaited(context.navigator.pushReplacement(
                      TopLevelPageRoute(builder: (_) => SignedInScreen())));
                }
              },
              initialUrl: services.get<AppConfig>().webUrl('login'),
            ),
```

File: lib/sign_in/widgets/form.dart [hunk 3]
```diff
@@ -82,48 +67,62 @@ class _SignInFormState extends State<SignInForm> {
...
+                    // in a non-hacky way in the future.
+                    await controller.evaluateJavascript('''
+                  var node = document.getElementById('loginarea');
+                  var html = document.getElementsByTagName('html')[0];
+                  html.removeChild(html.childNodes[2]);
+                  html.appendChild(node);
+                  document.getElementsByTagName('h2')[0].innerHTML = ''
+                  ''');
+                  } else if (url ==
+                      'https://${services.get<AppConfig>().domain}/dashboard') {
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: Can we already intercept this in `WebView.navigationDelegate`? Currently, we see a flash of the web-dashboard and are then forwarded to the app-dashboard.

File: lib/sign_in/widgets/form.dart [hunk 4]
```diff
@@ -82,48 +67,62 @@ class _SignInFormState extends State<SignInForm> {
...
+                    // Yes, this is not elegant. You may complain about it
+                    // when there is a nice way to get a single cookie via JavaScript.
+                    var jwt = cookies
+                        .split('; ')
+                        .firstWhere((element) => element.startsWith('"jwt='))
+                        .replaceAll('"', '')
+                        .substring(4);
+
+                    final storage = services.get<StorageService>();
+                    await storage.token.setValue(jwt);
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: Just storing the new JWT isn't enough. Have a look at `StorageService.setUserInfo` which also sets `userId` and `email`—we need to at least store `userId`, maybe the email address can be retrieved via the API

**Review by JonasWanke — CHANGES_REQUESTED** [review]
Dark theme isn't supported—custom CSS or wait for the web client?
→ ignore for now

File: lib/sign_in/widgets/form.dart (hunk 4 above)
Comment (janrenz) [reply lib/sign_in/widgets/form.dart]: i think there is a /me route

Comment (JonasWanke) [reply lib/sign_in/widgets/form.dart]: We can also retrieve it from the token like we did some time ago: https://github.com/schul-cloud/schulcloud-flutter/blob/3be9922ea19286fa6e4d855f8bb3f9208e190146/lib/app/services/storage.dart#L20-L28

File: lib/sign_in/widgets/form.dart [hunk 5]
```diff
@@ -1,10 +1,13 @@
 import 'package:black_hole_flutter/black_hole_flutter.dart';
 import 'package:flutter/material.dart';
+import 'package:flutter_inappwebview/flutter_inappwebview.dart';
 import 'package:flutter_svg/svg.dart';
 import 'package:pedantic/pedantic.dart';
 import 'package:schulcloud/app/app.dart';
+import 'package:schulcloud/sign_in/widgets/signin_browser.dart';
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: Please use a relative import for files of the same module

File: lib/sign_in/widgets/form.dart [hunk 6]
```diff
@@ -1,10 +1,13 @@
...
 import 'package:webview_flutter/webview_flutter.dart';
 
+import '../../app/app.dart';
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: Please use absolute imports for files of different modules. Also, the app module is already imported in line 6

File: lib/sign_in/widgets/signin_browser.dart [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import 'dart:convert';
+
+import 'package:flutter_inappwebview/flutter_inappwebview.dart';
+
+import '../../app/app.dart';
+import '../data.dart';
+
+class SignInBrowser extends InAppBrowser {
+  SignInBrowser({this.signedInCallback});
```
Comment (JonasWanke) [inline lib/sign_in/widgets/signin_browser.dart]:
```suggestion
  SignInBrowser({@required this.signedInCallback}) : assert(signedInCallback != null);
```

File: lib/sign_in/widgets/signin_browser.dart [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+  Function signedInCallback;
```
Comment (JonasWanke) [inline lib/sign_in/widgets/signin_browser.dart]:
```suggestion
  VoidCallback signedInCallback;
```

File: lib/sign_in/widgets/signin_browser.dart [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  @override
+  Future onLoadStart(String url) async {
+    final firstPathSegment = Uri.parse(url).pathSegments.first;
+    if (firstPathSegment == 'dashboard') {
+      var jwt = await CookieManager().getCookie(url: url, name: 'jwt');
+      await services.storage.token.setValue(jwt.value);
+      final rawResponse = await services.network.get('me');
+      final response = UserResponse.fromJson(json.decode(rawResponse.body));
+      await services.storage.setUserInfo(
+          email: response.email, userId: response.userId, token: jwt.value);
```
Comment (JonasWanke) [inline lib/sign_in/widgets/signin_browser.dart]: When arguments don't fit into a single line, prefer using a trailing comma to improve readability

File: lib/sign_in/widgets/signin_browser.dart [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+class SignInBrowser extends InAppBrowser {
+  SignInBrowser({this.signedInCallback});
+
+  Function signedInCallback;
+
+  @override
+  Future onLoadStart(String url) async {
+    final firstPathSegment = Uri.parse(url).pathSegments.first;
+    if (firstPathSegment == 'dashboard') {
+      var jwt = await CookieManager().getCookie(url: url, name: 'jwt');
```
Comment (JonasWanke) [inline lib/sign_in/widgets/signin_browser.dart]:
```suggestion
      final jwt = await CookieManager().getCookie(url: url, name: 'jwt');
```

File: lib/sign_in/widgets/signin_browser.dart [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+      await services.storage.token.setValue(jwt.value);
+      final rawResponse = await services.network.get('me');
+      final response = UserResponse.fromJson(json.decode(rawResponse.body));
+      await services.storage.setUserInfo(
+          email: response.email, userId: response.userId, token: jwt.value);
+      signedInCallback();
+      await close();
```
Comment (JonasWanke) [inline lib/sign_in/widgets/signin_browser.dart]: It looks like `storageManager.email` is never used. We could remove that preference and get all necessary information by decoding the returned JWT-token, saving one network call.

File: lib/sign_in/widgets/signin_browser.dart [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+import '../../app/app.dart';
+import '../data.dart';
+
+class SignInBrowser extends InAppBrowser {
+  SignInBrowser({this.signedInCallback});
+
+  Function signedInCallback;
+
+  @override
+  Future onLoadStart(String url) async {
```
Comment (JonasWanke) [inline lib/sign_in/widgets/signin_browser.dart]: Please add some logging to this method (see the old BLoC for an example)

File: lib/sign_in/widgets/signin_browser.dart [hunk 13]
```diff
@@ -0,0 +1,27 @@
+import 'dart:convert';
```
Comment (JonasWanke) [inline lib/sign_in/widgets/signin_browser.dart]: The file should be called `sign_in_browser.dart`

File: lib/sign_in/widgets/form.dart [hunk 14]
```diff
@@ -14,30 +17,26 @@ class SignInForm extends StatefulWidget {
...
-  String _ambientError;
+  @override
+  void initState() {
+    browser = SignInBrowser(signedInCallback: pushSignedInPage);
+    super.initState();
+  }
 
-  Future<void> _executeLogin(Future<void> Function() login) async {
-    setState(() => _isLoading = true);
+  void pushSignedInPage() {
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: Can be private

File: lib/sign_in/widgets/form.dart [hunk 15]
```diff
@@ -70,51 +69,17 @@ class _SignInFormState extends State<SignInForm> {
...
+                maxHeight: mediaQuery.size.height -
+                    400 -
+                    mediaQuery.padding.bottom -
+                    mediaQuery.padding.top,
+              ),
+              child: PrimaryButton(
+                  onPressed: () => browser.open(
+                      url: services.get<AppConfig>().webUrl('login'),
+                      options: InAppBrowserClassOptions()),
+                  child: Text(s.signIn_form_signIn))),
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: Please add trailing commas in multi-line widget creations. But now that we open the browser in a separate page, do we really need this `ConstrainedBox`?

File: lib/sign_in/widgets/form.dart [hunk 16]
```diff
@@ -70,51 +69,17 @@ class _SignInFormState extends State<SignInForm> {
...
-          ),
+              constraints: BoxConstraints(
+                maxHeight: mediaQuery.size.height -
+                    400 -
+                    mediaQuery.padding.bottom -
+                    mediaQuery.padding.top,
+              ),
+              child: PrimaryButton(
+                  onPressed: () => browser.open(
+                      url: services.get<AppConfig>().webUrl('login'),
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: `AppConfig` is available as `services.config`

**Review by JonasWanke — CHANGES_REQUESTED** [review]
Can you please rename remaining references of "login" to "signIn"?

Signing in doesn't work due to an NPE in `NetworkService`. Can you please change `...headers,` in `network.dart` line 220 to `if (headers != null) ...headers,`?

File: lib/sign_in/widgets/sign_in_browser.dart [hunk 17]
```diff
@@ -0,0 +1,32 @@
+import 'dart:convert';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter_inappwebview/flutter_inappwebview.dart';
+import 'package:meta/meta.dart';
+
+import '../../app/app.dart';
```
Comment (JonasWanke) [inline lib/sign_in/widgets/sign_in_browser.dart]: Please use absolute imports for files of different modules.

File: lib/sign_in/widgets/sign_in_browser.dart [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+  SignInBrowser({@required this.signedInCallback})
+      : assert(signedInCallback != null);
+
+  VoidCallback signedInCallback;
+
+  @override
+  Future onLoadStart(String url) async {
+    final firstPathSegment = Uri.parse(url).pathSegments.first;
+    if (firstPathSegment == 'dashboard') {
+      logger.i('Signing in...');
```
Comment (JonasWanke) [inline lib/sign_in/widgets/sign_in_browser.dart]:
```suggestion
      logger.i('Signing in…');
```

File: lib/sign_in/widgets/sign_in_browser.dart [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+      final jwt = await CookieManager().getCookie(url: url, name: 'jwt');
+      await services.storage.token.setValue(jwt.value);
+      final userIdJson = json
+          .decode(String.fromCharCodes(base64Decode(jwt.value.split('.')[1])));
+      await services.storage.userIdString.setValue(userIdJson['userId']);
+      logger.i('Signed in as userId ${userIdJson['userId']}');
```
Comment (JonasWanke) [inline lib/sign_in/widgets/sign_in_browser.dart]:
```suggestion
      logger.i('Signed in with userId ${userIdJson['userId']}');
```

File: lib/sign_in/utils.dart [hunk 20]
```diff
@@ -33,7 +33,6 @@ Future<bool> signOut(BuildContext context) async {
   if (confirmed) {
     // Actually log out.
 
-    await services.get<NetworkService>().delete('authentication');
```
Comment (JonasWanke) [inline lib/sign_in/utils.dart]: Why was this deleted?

File: lib/sign_in/widgets/form.dart [hunk 21]
```diff
@@ -61,35 +58,29 @@ class _SignInFormState extends State<SignInForm> {
...
-              child: PrimaryButton(
-                  onPressed: () => browser.open(
-                      url: services.get<AppConfig>().webUrl('login'),
-                      options: InAppBrowserClassOptions()),
-                  child: Text(s.signIn_form_signIn))),
+          PrimaryButton(
+              onPressed: () => browser.open(
+                  url: services.config.webUrl('login'),
+                  options: InAppBrowserClassOptions()),
+              child: Text(s.signIn_form_signIn)),
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: Please add trailing commas in multi-line widget creations.
```suggestion
          PrimaryButton(
            onPressed: () => browser.open(
                url: services.config.webUrl('login'),
                options: InAppBrowserClassOptions()),
            child: Text(s.signIn_form_signIn),
          ),
```

**Review by JonasWanke — CHANGES_REQUESTED** [review]
Can you please remove the `email`-field from `StorageService` as it's no longer populated in most cases? An then either remove `storageService.setUserInfo` or use it in `SignInBrowser` too.

While signing in works, I can no longer sign out. I still get back to the `SignInScreen`, but tapping "Sign in" I don't see the new WebView but get redirected to the dashboard immediately.

Also, please update this branch with master.

File: lib/sign_in/utils.dart (hunk 20 above)
Comment (ctiedt) [reply lib/sign_in/utils.dart]: This should have been `services.api.delete('authentication')` in the first place, but this results in a `401 Not authenticated` error. Trying the same request in Python or Postman, it works fine, I don't know what is different about Dart. However, it appears that even when it is successful, it does not change anything. Maybe someone with more API experience could shed some light on this?

File: lib/sign_in/utils.dart [hunk 22]
```diff
@@ -35,15 +37,10 @@ Future<bool> signOut(BuildContext context) async {
   if (confirmed) {
     // Actually log out.
 
-    unawaited(CookieManager().deleteAllCookies());
-    // This should probably be awaited, but right now awaiting it
-    // leads to the issue that logging out becomes impossible.
-    unawaited(services.storage.clear());
+    await CookieManager().deleteAllCookies();
+    await services.storage.clear();
```
Comment (JonasWanke) [inline lib/sign_in/utils.dart]: The actual signing out process is now happening in `SignOutScreen`. Please delete all cookies there; storage is already cleared:
https://github.com/schul-cloud/schulcloud-flutter/blob/28bc861a3334e3fb7d43e0d4463d685526147780/lib/sign_in/widgets/sign_out_screen.dart#L18-L23

File: lib/sign_in/utils.dart (hunk 20 above)
Comment (JonasWanke) [reply lib/sign_in/utils.dart]: After moving this to the `SignOutScreen` and calling this before clearing storage and cookies, it works for me.

File: lib/sign_in/widgets/form.dart [hunk 23]
```diff
@@ -56,21 +56,19 @@ class _SignInFormState extends State<SignInForm> {
...
           ),
           SizedBox(height: 32),
           PrimaryButton(
-              onPressed: () => browser.open(
-                  url: services.config.webUrl('login'),
-                  options: InAppBrowserClassOptions()),
-              child: Text(s.signIn_form_signIn)),
+            onPressed: () => browser.open(
+                url: services.config.webUrl('login'),
+                options: InAppBrowserClassOptions()),
```
Comment (JonasWanke) [inline lib/sign_in/widgets/form.dart]: What do you think about hiding the toolbar to not confuse the user?
```suggestion
            onPressed: () => browser.open(
              url: services.config.webUrl('login'),
              options: InAppBrowserClassOptions(
                inAppBrowserOptions: InAppBrowserOptions(toolbarTop: false),
              ),
            ),
```

**Review by JonasWanke — CHANGES_REQUESTED** [review]
Looks good! Only some minor changes left, and then we can get this merged!

**Review by JonasWanke — APPROVED** [review]
